### ISSUE TITLE
How to configure the api key to be mandatory?

### ISSUE THREAD

**Issue by hanssens** [mihirdilip/aspnetcore-authentication-apikey#3]
I'm encountering the scenario where if you don't add a header key, it will skip authentication. 

For example, this works - as it will nicely trigger the middleware:

```
curl --location --request GET 'http://localhost:5000?Authorization=w00t'
```

However, **skipping** or not providing a key will simply skip the authentication altogether:

```
curl --location --request GET 'http://localhost:5000'
```

Is there a way to make the api key mandatory? 

Edit: This issue has also been noted [in this comment](https://github.com/mihirdilip/aspnetcore-authentication-apikey/issues/2#issuecomment-591697835), but not followed-up (_"... the ApiKeyProvider will not be called if the expected apikey header is not present"_).

**Comment by hanssens** [comment]
Here is a reproducible sample, as I'm using it with a clean, skeleton AspNetCore 3.1 API, with the following two modifications.

Example of **Startup.cs**:

```
    public class Startup
    {
        // This method gets called by the runtime. Use this method to add services to the container.
        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddApiKeyAuthentication("My Test Api");
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UseRouting();
            app.UseApiKeyAuthentication();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapGet("/", async context => { await context.Response.WriteAsync("Hello World!"); });
            });
        }
    }
```

And example of my custom middleware, grouped in **ApiKeyAuthentication.cs**:

```
    public static class ApiKeyAuthentication
    {
        public static void AddApiKeyAuthentication(this IServiceCollection services, string apiName)
        {
            // AddApiKeyInHeaderOrQueryParams extension takes an implementation of IApiKeyProvider for validating the key. 
            // It also requires Realm and KeyName to be set in the options.
            services.AddAuthentication(ApiKeyDefaults.AuthenticationScheme)
                
                // allow API key in both header + as query param
                .AddApiKeyInHeaderOrQueryParams<PoorManValidationService>(options => 
                { 
                    options.Realm = apiName; 
                    options.KeyName = "Authorization";
                });
        }

        public static void UseApiKeyAuthentication(this IApplicationBuilder app)
        {
            app.UseAuthentication();
        }
    }
    
    public class PoorManValidationService : IApiKeyProvider
    {
        public async Task<IApiKey> ProvideAsync(string key)
        {
            if (string.IsNullOrEmpty(key)) throw new ApiKeyAuthenticationException("Missing 'Authorization' token in either header or as query parameter.");

            // For the sake of testing, key must be `w00t`
            if (key.Equals("w00t", StringComparison.InvariantCultureIgnoreCase))
            {
                var response = new ApiKey(key: key, owner: "owner", new List<Claim>());
                return await Task.FromResult(response);
            }
            
            // Log failures here through ILogger<PoorMansValidationService> logger?
            throw new ApiKeyAuthenticationException("Unknown or invalid authorization token provided.");
        }
    }
    
    class ApiKey : IApiKey
    {
        public string Key { get; }
        public string OwnerName { get; }
        public IReadOnlyCollection<Claim> Claims { get; }
        
        public ApiKey(string key, string owner, List<Claim> claims = null)
        {
            Key = key;
            OwnerName = owner;
            Claims = claims ?? new List<Claim>();
        }
    }

}

 public class ApiKeyAuthenticationException : Exception
 {
     public ApiKeyAuthenticationException(string message) : base(message) { }
 }
````

**Comment by mihirdilip** [comment]
@hanssens, it is not the responsibility of this library to make every endpoint on your server to be authorized. It is part of ASP.NET Core and it is upto you how you want to handle it. 

Try the below..

```
    public class Startup
    {
        // This method gets called by the runtime. Use this method to add services to the container.
        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
        public void ConfigureServices(IServiceCollection services)
        {
            // For MVC you can use this to apply the authorization to every call.
            //services.AddMvc(options =>
            //{
            //    options.Filters.Add(new AuthorizeFilter(new AuthorizationPolicyBuilder().RequireAuthenticatedUser().Build()));
            //});

            // In your case, use the below
            services.AddAuthorization();
            services.AddAuthentication(ApiKeyDefaults.AuthenticationScheme)

                // allow API key in both header + as query param
                .AddApiKeyInHeaderOrQueryParams<PoorManValidationService>(options =>
                {
                    options.Realm = "My Test Api";
                    options.KeyName = "Authorization";
                });
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UseRouting();

            // NOTE THIS HERE!!!!
            app.UseAuthentication();
            app.UseAuthorization();

            app.UseEndpoints(endpoints =>
            {
                // For MVC you can use this 
                // endpoints.MapDefaultControllerRoute();

                // In your case, use the below
                endpoints.MapGet("/", async context =>
                {
                    await context.Response.WriteAsync("Hello World!");
                }).RequireAuthorization();  // NOTE THIS HERE!!!! 
            });
        }
    }
```
I hope this helps.

Let me know if you have any issues.

**Comment by hanssens** [comment]
Thanks @mihirdilip. I understand your concern. However, would a feature that makes this behaviour configurable make sense?

**TL;DR -** I've setup [a simple change](https://github.com/hanssens/aspnetcore-authentication-apikey/commit/0f2b55baab73474e5933de83076b4e8c39eeda44) (work in progress, but you get the idea) which I think would add value to your library. What's your opinion on making this behaviour configurable + would you be willing to accept a PR on it?

**Comment by mihirdilip** [comment]
Thanks for your feedback @hanssens, I am sorry I am not currently accepting any PR. I will however keep this in mind for any future updates.

For your solution all you have to do is chain .RequireAuthorization() in your implemetation for the authorization to work.

```
app.UseEndpoints(endpoints =>
            {
                endpoints.MapGet("/", async context =>
                {
                    await context.Response.WriteAsync("Hello World!");
                }).RequireAuthorization();  // NOTE THIS HERE!!!! 
            });
```


And regarding [this comment](https://github.com/mihirdilip/aspnetcore-authentication-apikey/issues/2#issuecomment-591697835), I must have misunderstood the comment. I thought that @farlee2121 got it working and so I closed the issue.

**Comment by hanssens** [comment]
Cheers. That works indeed. But there was an underlying problem at my end why the [Authorize] attribute wasn't challenged. 

1. Accidently `System.Web.Http.Authorize` was used to decorate certain Controllers, instead of `Microsoft.AspNetCore.Authorization`. 
2. ApiKeyAuthentication needs to be registered _before_  `app.UseAuthorization()`

For me, this closes the issue. 

This means that the 'IsRequired' suggestion indeed is not the way to go. But as a friendly suggestion, perhaps you might want to add something on the documentation on this?

> N.B. Too bad you're not accepting any PR's. But I respect it, of course. Should you ever change your mind, for code or documentation, I'm sure people are willing to contribute.

**Comment by mihirdilip** [comment]
Thanks for you feedback. I will add some notes to the documentation soon.

Yes with netcore 3.0 and later you have to use it in a specific order.

```
            app.UseRouting();

            // NOTE THIS HERE!!!!
            app.UseAuthentication();
            app.UseAuthorization();

            app.UseEndpoints
```

If you are using controllers, you can perhaps always challenge all the requests by adding either of below on Startup class (depending what you are using)

```
            services.AddMvc(options => {
                options.Filters.Add(new AuthorizeFilter(new AuthorizationPolicyBuilder().RequireAuthenticatedUser().Build()));
            });
```
```
            services.AddControllers(options => { 
                options.Filters.Add(new AuthorizeFilter(new AuthorizationPolicyBuilder().RequireAuthenticatedUser().Build()));
            });
```
